[E2E][QIT] Migrate Github Actions workflows to work with QIT

mgascam opened the pull request:
Fixes #WOOPMNT-5365

This PR migrates the existing E2E GitHub workflows to use QIT, completing the workflow infrastructure portion of the QIT E2E migration effort. This is part of a stacked PR series where each PR builds incrementally toward full QIT adoption.

#### Current State & Scope

🚧 Important: This PR focuses on infrastructure migration - moving from legacy E2E workflows to QIT-based workflows. The test suite is functionally complete and running in CI, but stability work is needed as a follow-up task (see [WOOPMNT-5361: \[E2E\]\[QIT\] Fix timeout errors and flaky specs in CI](https://linear.app/a8c/issue/WOOPMNT-5361/e2eqit-fix-timeout-errors-and-flaky-specs-in-ci)).

The intent is to get the foundation merged and reviewed without blocking on every flaky test. This should allow parallel work on stabilization while infrastructure is in place and track infrastructure changes separately from test fixes.

⚠️ Current Limitations (to be addressed in follow-up work):
1. Test Flakiness: Some tests have timeout issues and intermittent failures in CI
2. Retry Logic: While automatic retries help, some specs may need additional stability work

#### Changes proposed in this Pull Request

This PR establishes the QIT E2E testing infrastructure with:

1. Separated Workflows for Different Use Cases:
    - `qit-e2e-pull-request.yml`: Focused PR testing (10 jobs: L-1 + stable WC versions across 5 test categories)
    - `qit-e2e.yml `(renamed to "Scheduled & Push"): Comprehensive version matrix (25+ jobs including legacy WC 7.7.0, beta, RC, WP nightly)
2. Dynamic Version Resolution:
    - New script: `.github/scripts/generate-qit-wc-matrix.sh`
    - Automatically fetches L-1 WooCommerce version from WordPress.org API
    - Checks availability of beta/RC versions dynamically
3. Build Once, Test Many:
    - New build-plugin job: Builds `woocommerce-payments.zip` once and uploads as artifact
    - All matrix jobs download the pre-built artifact instead of rebuilding
    - Impact: Saves ~5 minutes × number of matrix jobs (e.g., 50 minutes saved on 10-job PR runs)
    - Ensures all tests run against the identical build artifact
4. Test Execution Features:
    - Tag-based filtering: Single tags (e.g., `merchant`) auto-exclude subscriptions/blocks; multiple tags use AND logic
    - Automatic retries: Failed specs are parsed from CTRF reports and retried once
    - WordPress version support: Can test against stable, nightly, and RC WordPress builds
    - Manual dispatch: Run workflows with custom WC/PHP/WP versions and specific test tags
5. Error Handling:
    - Validates QIT results directory exists before proceeding
    - Parses CTRF format reports to identify failed specs
    - Provides clear error messages for QIT setup failures
    - Uses --no_upload_report to keep CI results internal

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

What Should Work:
  - Workflows trigger on appropriate events (PR, push, schedule)
  - Matrix generates correctly with dynamic versions
  - Tests execute via QIT
  - Failed specs are identified and retried
  - Results are logged (not uploaded to QIT)

What May Not Work (expected, to be fixed later):
  - Some tests may timeout or fail intermittently
  - Retry attempts may not always succeed
  - Full test suite will not be 100% green

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

mgascam committed:
Update QIT E2E workflow to enhance testing matrix and add test tag input

mgascam committed:
Refactor QIT test execution path and configuration for improved clarity

mgascam committed:
Refactor QIT test tag handling to remove leading '@' symbol and improve tag processing logic

mgascam committed:
Enhance QIT E2E workflow to include WordPress version options and expand test matrix for pull requests

mgascam committed:
Refactor QIT E2E test matrix to use space-separated tagg

mgascam committed:
Refactor QIT E2E test matrix to exclude specific tags for merchant and shopper jobs to avoid duplication

mgascam committed:
Refactor QIT E2E test matrix to use space-separated tags for merchant and shopper jobs, removing exclusion tags to avoid duplication. Remove retries

mgascam committed:
Enhance QIT E2E workflow to include first run and retry logic for failed specs, allowing for better error handling and reporting.

mgascam committed:
Refactor QIT E2E workflow to improve test matrix generation and streamline test execution paths, enhancing clarity and maintainability.

mgascam committed:
Temporarily disable develop and trunk branch runs in QIT E2E workflow to reduce CI load

mgascam committed:
Refactor QIT E2E workflow to comment out specific matrix entries and improve test directory determination logic for better clarity and maintainability.

mgascam committed:
Refactor QIT E2E workflow to streamline test selection logic and disable retries, enhancing clarity and consistency in test execution.

mgascam committed:
Refactor QIT E2E workflow to enhance test selection logic using tags, improving clarity and flexibility in test execution.

mgascam committed:
Fix grep-invert

mgascam committed:
Fix failed specs extraction

mgascam committed:
Disable retries for failed specs

mgascam committed:
chore: tickle PR view

mgascam committed:
Fix grep-invert error

mgascam committed:
Refactor QIT test retries to use grep pattern for failed specs

mgascam committed:
Refactor QIT results handling to improve error messages and file search order

mgascam committed:
Remove branch reference for QIT E2E workflow enhancements to streamline CI configuration

mgascam committed:
Enhance QIT test execution with error handling and result verification

mgascam committed:
Refactor QIT E2E workflow to separate pull request and scheduled runs

mgascam committed:
Restore QIT download and authentication

mgascam committed:
Update QIT E2E workflow to use stable version for compatibility testing

mgascam committed:
Add QIT script for WooCommerce version matrix and update workflows to use it

mgascam committed:
Add --no_upload_report option to QIT E2E tests in CI

mgascam committed:
Merge branch 'dev/qit-e2e-subscriptions-specs-migration' into dev/qit-e2e-enhance-workflows

adimoldovan commented on .github/workflows/qit-e2e-pull-request.yml:
```diff
@@ -0,0 +1,350 @@
+        run: |
+          RESULTS_JSON="${{ steps.parse_failed_specs.outputs.results_json }}"
+
+          # Extract failed spec file paths from CTRF format
```
Why not using [Playwright's last-failed builtin feature](https://playwright.dev/docs/running-tests#run-last-failed-tests)? It should greatly simplify the logic, no need to parse any reports.

adimoldovan commented on .github/workflows/qit-e2e-pull-request.yml:
```diff
@@ -0,0 +1,350 @@
+        required: false
+        default: ''
+        type: string
+      run_ui_mode:
```
What is the benefit of running in UI mode in CI?

adimoldovan commented on .github/workflows/qit-e2e-pull-request.yml:
```diff
@@ -0,0 +1,350 @@
+
+      - name: First Run QIT Tests
+        id: first_run_qit_tests
+        if: ${{ github.event_name != 'pull_request' || github.event.pull_request.head.repo.full_name == github.repository }}
```
Shouldn't this be `== 'pull_request'`?

adimoldovan commented on .github/workflows/qit-e2e-pull-request.yml:
```diff
@@ -0,0 +1,350 @@
+            if [[ "$WORD_COUNT" -eq 1 ]]; then
+              # Single tag: exclude subscriptions and blocks tests using --grep-invert with regex alternation
+              # Example: --grep @shopper --grep-invert "@(subscriptions|blocks)"
+              OPTIONS="${OPTIONS} --pw_test_tag=\"${CLEAN_TAG}\" --pw_options=\"--retries=0 --grep-invert '@(subscriptions|blocks)'\""
```
I don't understand the need of this step or how it works. What if I use the tag `@subscriptions` or `@blocks`?

adimoldovan commented on .github/workflows/qit-e2e-pull-request.yml:
```diff
@@ -0,0 +1,350 @@
+          # Check if QIT actually ran tests by looking for results directory
+          QIT_RESULTS_DIR=$(find /tmp -maxdepth 1 -name "qit-results-*" -type d 2>/dev/null | sort -r | head -1)
+          if [[ -z "$QIT_RESULTS_DIR" ]]; then
+            echo "::error::QIT did not produce test results. This indicates an environment or setup failure."
```
Would QIT exit with code 0 if the setup failed?

adimoldovan reviewed (commented):
Nice to see this moving forward @mgascam!

> 1. Separated Workflows for Different Use Cases:

From what I can see there's not much difference between the 2 workflows, most is duplicated logic (and also quite verbose). I think a single workflow would be better here.

> Tag-based filtering: Single tags (e.g., merchant) auto-exclude subscriptions/blocks; multiple tags use AND logic

How much flexibility do you expect you'd need to run these tests manually in GitHub? I think organizing the tests in [projects](https://playwright.dev/docs/test-projects#splitting-tests-into-projects) would be easier than dealing with parsing tags in the CI workflow. And use project name(s) as input in the workflow.

mgascam committed:
Add Playwright test results upload step to E2E workflows

adimoldovan commented:
@mgascam is this still on track? Let me know if I can help.

mgascam commented:
> @mgascam is this still on track? Let me know if I can help.

Hi @adimoldovan, apologies for the delayed reply. I’ve been juggling a few other priorities and had to pause this work. I’ll pick it back up next week, starting with your comments.

Just to share some context: I realized some of the previous PRs were too big and likely hard to review, so I’m restructuring the work into smaller, more approachable chunks. I’ve started with the shopper tests (https://github.com/Automattic/woocommerce-payments/pull/11133) to make progress smoother for everyone.

This shouldn’t affect this PR much, I don’t expect any fundamental changes here. I’ll continue addressing your feedback as I move forward.

mgascam commented:
Closing in favor of https://github.com/Automattic/woocommerce-payments/pull/11277
